File: ex42/ex42.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
public:
    int trap(vector<int>& height)
    {
        int n = height.size();
        int l = 0, r = n - 1;
        int lmax = 0, rmax = 0;
        int ans = 0;
        while (l < r)
        {
            lmax = max(height[l], lmax);
            rmax = max(height[r], rmax);
            if (height[l] < height[r])
                ans += lmax - height[l++];
            else
                ans += rmax - height[r--];
        }
        return ans;
    }
};

int main(void)
{

    return 0;
}

File: ex42/ex42.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        l = 0
        r = n - 1
        ans = 0
        lmax = rmax = 0
        while l < r:
            lmax = max(lmax, height[l])
            rmax = max(rmax, height[r])
            if height[l] < height[r]:
                ans += lmax - height[l]
                l += 1
            else:
                ans += rmax - height[r]
                r -= 1
        return ans